specs/extension_image.md

```diff
@@ -35,8 +35,109 @@ confext images extend `/etc`. They must contain a `/etc/extension-release.d/exte
+
+## Base Directory Immutability / Mutability
+The contents of system and configuration extension images are immutable and cannot be changed after
+the image was created.
+By default, the base directory (`/etc`, `/usr`, `/opt`) overlaid by an extension is also rendered immutable
+when the extension is applied ("merged").
+This only affects base directories actually contained in extensions merged to a root filesystem:
+for instance, if a sysext extends `/usr` but not `/opt`, `/opt` will remain mutable.
+However, immutability affects the _full_ base directory.
+An extension shipping a single custom path e.g. below `/etc/appconfig-extra/some/sub/path` will still
+render the entirety of `/etc` immutable.
+The base directory or directories remain immutable for as long as the extensions are merged.
+Original state is regained only after all extensions overlaying a base directory were un-merged.
+
+By implication, merging a confext will result in `/etc` becoming read-only, and merging a sysext might
+render `/usr` and/or `/opt` read-only, depending on the sysext's contents.
+
+### Optional Mutability
+While overlaid base directories are immutable by default, implementations may provide options
+for retaining mutability.
+Retaining mutability may for instance be useful for compatibility with general purpose applications,
+enabling users to operate a "mixed mode" with both system / configuration extensions
+and traditional applications.
+Mixed mode may also be integrated by distributions to facilitate a smooth transition from
+traditional package management to a purely image-based composition of the root file system
+if taking such a direction is desired.
+
+### Extension Overlay [Im]Mutability Modes
+System and configuration extensions may operate in one of three modes.
+
+1. *Immutable mode* - The overlaid base directory is rendered immutable.
+    This is the default.
+    Files and paths in extensions will by implication override files and paths also present
+    in the base directory.
+2. *Mutable mode* - Writes are directed to an _upperdir_ specified by the user / operator
+    (see "Mutability Mode Configuration" below).
+    This _upperdir_ will contain all changes made to the overlaid base directory.
+    It also changes the stacking order compared to immutable mode in that files and paths present
+    in _upperdir_ will override files and paths that are also present in an extension.
+    1. _Upperdir_ may be specified to _be_ the base directory: may be used to retain mutability
+       of the base directory after extensions were merged.
+    2. Alternatively, _upperdir_ may be an entirely separate directory: modifications will be captured
+       but the base directory will remain unchanged, retaining its state from before the extension was merged.
+3. *Ephemeral Mode* - Similar to mutable mode (2.) above but writes are explicitly stored
+    in a temporary directory (possibly in RAM) and discarded as soon as extensions are un-merged.
```

Comment:
Tmpfs is swappable, so it's not actually "RAM". I think it's better to say "in a temporary file system" and nothing more.

Reply:
I'll change it to "will only be stored temporarily while sysexts are merged". I'd like to avoid implementation details and I feel my phrasing above is too specific already.